src/framework/MOM_random.F90

```diff
@@ -0,0 +1,462 @@
+  real, dimension(SZI_(HI),SZJ_(HI)), intent(out) :: rand !< Random numbers between 0 and 1
+  ! Local variables
+  integer :: i,j
+
+  do j = HI%jsd,HI%jed
+    do i = HI%isd,HI%ied
+      rand(i,j) = getRandomReal( CS%stream2d(i,j) )
+    enddo
+  enddo
+
+end subroutine random_2d_01
+
+!> Returns an approximately normally distributed random number with mean 0 and variance 1
+!! for each cell of the model grid
+subroutine random_2d_norm(CS, HI, rand)
+  type(PRNG),           intent(inout) :: CS !< Container for pseudo-random number generators
+  type(hor_index_type), intent(in)    :: HI !< Horizontal index structure
+  real, dimension(SZI_(HI),SZJ_(HI)), intent(out) :: rand !< Random numbers between 0 and 1
+  ! Local variables
+  integer :: i,j,n
+
+  do j = HI%jsd,HI%jed
+    do i = HI%isd,HI%ied
+      rand(i,j) = getRandomReal( CS%stream2d(i,j) ) - 0.5
+    enddo
+    do n = 1,11
+      do i = HI%isd,HI%ied
+        rand(i,j) = rand(i,j) + ( getRandomReal( CS%stream2d(i,j) ) - 0.5 )
+      enddo
+    enddo
+  enddo
+
+end subroutine random_2d_norm
+
+!> Constructor for scalar PRNG. Can be used to reset the sequence.
+subroutine random_0d_constructor(CS, Time, seed)
+  type(PRNG),      intent(inout) :: CS   !< Container for pseudo-random number generators
+  type(time_type), intent(in)    :: Time !< Current model time
+  integer,         intent(in)    :: seed !< Seed for PRNG
+  ! Local variables
+  integer :: tseed
+
+  tseed = seed_from_time(Time)
+  tseed = ieor(tseed, seed)
+  CS%stream0d = new_RandomNumberSequence(tseed)
+
+end subroutine random_0d_constructor
+
+!> Constructor for gridded PRNG. Can be used to reset the sequence.
+subroutine random_2d_constructor(CS, HI, Time, seed)
+  type(PRNG),           intent(inout) :: CS   !< Container for pseudo-random number generators
+  type(hor_index_type), intent(in)    :: HI   !< Horizontal index structure
+  type(time_type),      intent(in)    :: Time !< Current model time
+  integer,              intent(in)    :: seed !< Seed for PRNG
+  ! Local variables
+  integer :: i,j,sseed,tseed
+
+  if (.not. allocated(CS%stream2d)) allocate( CS%stream2d(HI%isd:HI%ied,HI%jsd:HI%jed) )
+
+  tseed = seed_from_time(Time)
+  tseed = ieor(tseed*9007, seed)
+  do j = HI%jsd,HI%jed
+    do i = HI%isd,HI%ied
+      sseed = seed_from_index(HI, i, j)
+      sseed = ieor(tseed, sseed*7993)
+      CS%stream2d(i,j) = new_RandomNumberSequence(sseed)
+    enddo
+  enddo
+
+end subroutine random_2d_constructor
+
+!> Return a seed derived as hash of values in Time
+integer function seed_from_time(Time)
+  type(time_type), intent(in)    :: Time !< Current model time
+  ! Local variables
+  integer :: yr,mo,dy,hr,mn,sc,s1,s2
+
+  call get_date(Time,yr,mo,dy,hr,mn,sc)
+  s1 = sc + 61*(mn + 61*hr) + 379 ! Range 379 .. 89620
+  ! Fun fact: 2147483647 is the eighth Mersenne prime.
+  ! This is not the reason for using 2147483647+1 here.
+  s2 = mod(dy + 32*(mo + 13*yr), 2147483648) ! Range 0 .. 2147483647
```

Comment:
I am not able to build with gnu/8.3.0 because of the following error:

Error: Integer too big for its kind at (1).

Should it be 2147483647 (default value of HUGE for 32-bit integers, see post below) instead of 2147483648?

https://stackoverflow.com/questions/34972531/gfortran-error-integer-too-big-for-its-kind-at-1?rq=1

Reply:
I agree, this should have at least raised a warning somewhere.

I'm more confused that no one else (including me) is having the same problem.  Many of us need to run at 32-bit integers to keep compatibility with our libraries (MPI, etc).

Reply:
Is the mod required?  2147483648/(32*13) > 5M (years)

Reply:
@gustavo-marques Does using `mod(dy + 32*(mo + 13*yr), 2147483647)` work with gnu/8.3.0? I think the mod is required for negative years.

Reply:
If you're dealing with negative years, don't you need to use `modulo` rather than `mod`?

Reply:
Wouldn't this work as intended if we just changed 2147483648 to 2147483648_8? If the MOD intrinsic not support mixed kind integer arguments for some compilers, we could also add the declaration `kind=8` to the declaration of dy or one of the other variables.

Reply:
Just a FYI, kind=8 is not always 8bytes (e.g. NAG).

Reply:
- The number was chosen to fit into 4-bytes. I believe all our integers are 4-byte integers and not only due to a limitation in netCDF. I made the mistake of assuming integers were 8-bytes [before](https://github.com/NOAA-GFDL/icebergs/commit/0348755c90bcfac0bd444dc97be1168a7b72d5c8) and think I was avoiding that mistake here.
- The intent was create a positive integer so the `mod` should indeed be a `modulo`
- I'm trying to find somewhere I have access to gcc-8.3.0 to test a patch.

Reply:
> @gustavo-marques Does using `mod(dy + 32*(mo + 13*yr), 2147483647)` work with gnu/8.3.0? I think the mod is required for negative years.

Yes, it does.

Reply:
Excellent. I'm sending a patch.

Reply:
#1113 should fix this problem. We'll merge it into this PR if @gustavo-marques can confirm it works.